bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/AbstractResourceIconProvider.java

```diff
@@ -0,0 +1,110 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.ui.icon;
+
+import java.io.InputStream;
+import java.util.Set;
+
+import org.eclipse.smarthome.ui.icon.IconSet.Format;
+import org.osgi.framework.BundleContext;
+
+/**
+ * This is an abstract base class for implementing icon providers that serve icons from file resources.
+ * These files could be read from the file system, directly from the bundle itself or from somewhere else that can
+ * provide an {@link InputStream}.
+ *
+ * The resources are expected to follow the naming convention "<category>[-<state>].<format>", e.g. "alarm.png" or
+ * "alarm-on.svg".
+ * Resource names must be all lower case. Whether an icon is provided or not is determined by the existence of a
+ * resource without a state postfix.
+ * If a specific resource for a state is available, it will be used. If not, the default icon without a state postfix is
+ * used. If the state is a decimal number between 0 and 100, the implementation will look for a resource with the next
+ * smaller state postfix available. Example: For category "DimmableLight" and state 84, it will check for the resources
+ * dimmablelight-82.png, dimmablelight-81.png, dimmablelight-80.png and return the first one it can find.
+ *
+ * @author Kai Kreuzer
+ *
+ */
+abstract public class AbstractResourceIconProvider implements IconProvider {
+
+    /**
+     * The OSGi bundle context
+     */
+    protected BundleContext context;
+
+    /**
+     * When activating the service, we need to keep the bundle context.
+     *
+     * @param context the bundle context provided through OSGi DS.
+     */
+    protected void activate(BundleContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public Set<IconSet> getIconSets() {
+        return getIconSets(null);
+    }
+
+    @Override
+    public Integer hasIcon(String category, String iconSetId, Format format) {
+        return hasResource(iconSetId, category.toLowerCase() + "." + format.toString().toLowerCase()) ? getPriority()
+                : null;
+    }
+
+    @Override
+    public InputStream getIcon(String category, String iconSetId, String state, Format format) {
+        String resourceWithState = category + ((state != null) ? "-" + state.toLowerCase() : "") + "."
+                + format.toString().toLowerCase();
+        if (hasResource(iconSetId, resourceWithState)) {
+            return getResource(iconSetId, resourceWithState);
+        } else {
+            // let's treat all percentage-based categories
+            try {
+                Double stateAsDouble = Double.valueOf(state);
+                if (stateAsDouble > 0 && stateAsDouble < 100) {
+                    for (int i = stateAsDouble.intValue(); i >= 0; i--) {
+                        String resourceWithNumberState = category + "-" + i + "." + format.toString().toLowerCase();
+                        if (hasResource(iconSetId, resourceWithNumberState)) {
```

Comment:
In worst case if the state is 99 it tries to look up 99 files? And why looking for the smaller resource? If there are Icons for 0, 25, 75 and 100, wouldn´t you expect that it returns the Icon for 75 if the status is 74?